Handle DisplayAttribute.

adamradocz opened the pull request:
Closes #11 

Handles the `[Display(Name = "Customer name")]` attribute. It works on .NET Core 3.1+, but unfortunately, I couldn't figure out why doesn't it work on .NET Framework 4.8.

andrewlock commented on src/NetEscapades.EnumGenerators/EnumToGenerate.cs:
```diff
@@ -8,15 +8,19 @@ namespace NetEscapades.EnumGenerators;
     public readonly bool HasFlags;
     public readonly string UnderlyingType;
-    public readonly List<KeyValuePair<string, object>> Values;
+
+    /// <summary>
+    /// Key is the enum name. Value is the custom name setted by the <c>[Display(Name)]</c> attribute.
+    /// </summary>
+    public readonly List<KeyValuePair<string, string?>> Values;
```
I think we should give this a different name? It's not the "values" now, it's the `Names`?

adamradocz replied:
Yes, ´Names´ much better.

andrewlock commented on tests/NetEscapades.EnumGenerators.Tests/EnumGeneratorTests.cs:
```diff
@@ -134,4 +134,33 @@ internal enum MyEnum
         return Verifier.Verify(output).UseDirectory("Snapshots");
     }
+
+#if NETCOREAPP3_1_OR_GREATER
```
We definitely need to figure out why this doesn't work 🤔 Maybe it just doesn't work in the unit tests, perhaps because we're missing an assembly reference?

adamradocz replied:
I have no clue what else should be needed besides `System.ComponentModel.DataAnnotations`

andrewlock replied:
Me neither 😂 The question is whether this just doesn't work in the unit tests and works fine in the integration tests. My suspicion is that we need additional references [here](https://github.com/andrewlock/NetEscapades.EnumGenerators/blob/main/tests/NetEscapades.EnumGenerators.Tests/TestHelpers.cs#L21) to make it work, but writing an integration test would be a good bet 🙂

adamradocz replied:
When I was debugging that's what I noticed:
The Generator can't see the `Name` values of the `Display` attribute, when the target framework is .NET 4.8.
But, when I switch to .NET Core, the values was present.

adamradocz replied:
Added the integration test.

andrewlock reviewed (commented):
This is good work, thanks! Before we merge it, I think we should do 2 things:

1. Add an integration test for the behaviour
2. Figure out why it doesn't work in .NET Framework

Definitely happy to merge it once we have those, thanks again!

FroggieFrog commented:
If you add a reference to the correct NuGet package and load the corresponding library "it just works".

Following steps I did:

1. Add package `System.ComponentModel.Annotations` to `NetEscapades.EnumGenerators.Tests` and `NetEscapades.EnumGenerators`.
2. Create Instance of `DisplayAttribute` inside the test to load the assembly (I don't know a better way).
3. It works.

### Before

.Net 4.8 without PackageReference (also notice the different attributeName)

![net48_without](https://user-images.githubusercontent.com/6222752/161263473-0cc9e999-fb80-48ca-bb66-2d8822f70eb0.png)

### After

.Net 4.8 without PackageReference

![net48_with](https://user-images.githubusercontent.com/6222752/161263532-2731e1ed-5fc1-4c5b-b4d2-747efd0536f4.png)

#### some code

```csharp
[Fact]
public Task CanGenerateEnumExtensionsWithDisplayName()
{
    var x = new System.ComponentModel.DataAnnotations.DisplayAttribute();
    //const string input = ...
    //...
}
```

adamradocz commented:
@FroggieFrog it seems you put a lot of work on this. Don't you wanna update the PR?

adamradocz commented:
> This is good work, thanks! Before we merge it, I think we should do 2 things:
> 
> 1. Add an integration test for the behaviour
> 2. Figure out why it doesn't work in .NET Framework
> 
> Definitely happy to merge it once we have those, thanks again!

Done.

andrewlock commented on tests/NetEscapades.EnumGenerators.IntegrationTests/ExtensionTests.cs:
```diff
@@ -14,8 +17,23 @@ public abstract class ExtensionTests<T> where T : struct
+        {
+            if (valueAsString is not null)
+            {// Prevent: Warning CS8604  Possible null reference argument for parameter 'name' in 'MemberInfo[] Type.GetMember(string name)'
+                var memberInfo = value.GetType().GetMember(valueAsString);
```
nit: not a blocker though!
```suggestion
                var memberInfo = typeof(value).GetMember(valueAsString);
```

adamradocz replied:
Actually `typeof(T)`. Fix it.

andrewlock commented on src/NetEscapades.EnumGenerators/NetEscapades.EnumGenerators.csproj:
```diff
@@ -13,11 +13,12 @@
   <ItemGroup>
     <PackageReference Include="Microsoft.CodeAnalysis.Analyzers" Version="3.3.2" PrivateAssets="all" />
     <PackageReference Include="Microsoft.CodeAnalysis.CSharp" Version="4.0.1" PrivateAssets="all" />
+    <PackageReference Include="System.ComponentModel.Annotations" Version="5.0.0" Condition="'$(TargetFramework)' == 'net48'" />
```
I don't think we should add this to the Source generator project itself? 🤔 I think we can probably get away with only adding it to the test projects - that's what users would need to do to use the attribute I believe? 

```suggestion
```

adamradocz replied:
You're right. Fixed.

andrewlock commented on tests/NetEscapades.EnumGenerators.IntegrationTests/EnumWithDisplayNameInNamespaceExtensionsTests.cs:
```diff
@@ -0,0 +1,60 @@
+        (EnumWithDisplayNameInNamespace)3,
+    };
+
+    public static TheoryData<string> ValuesToParse() => new()
+    {
+        "First",
+        "Second",
+        "first",
+        "SECOND",
+        "3",
+        "267",
+        "-267",
+        "2147483647",
+        "3000000000",
+        "Fourth",
+        "Fifth",
+    };
```
Thanks for adding this test! This has made me wonder if we should also allow _parsing_ via the `[Display]` attribute too 🤔 Otherwise the values aren't round-trippable. I'm also wondering if we should make using the attributes round-trippable

adamradocz replied:
Actually, I need that function, so I was going to suggest the same.

What do you mean by **"attributes round-trippable"**?

adamradocz replied:
Done.

andrewlock reviewed (commented):
Thanks @adamradocz, this is looking really good! 😃 Just a couple of nits, but I'm wondering if we should include the `[Display]` name in the `TryParse()` etc methods too 🤔 I'm undecided, what do you think? Also wondering if we should make this attribute functionality optional, but we could always add that later if needs be!

adamradocz commented:
- Now the Display attribute is parsed back.
- Refactored the `TryParse` a little bit, so only the private method contains the logic.
- Added a few more tests.

adamradocz commented:
``` ini

BenchmarkDotNet=v0.13.1, OS=Windows 10.0.22000
11th Gen Intel Core i7-11800H 2.30GHz, 1 CPU, 16 logical and 8 physical cores
  [Host]     : .NET Framework 4.8 (4.8.4470.0), X64 RyuJIT  [AttachedDebugger]
  DefaultJob : .NET Framework 4.8 (4.8.4470.0), X64 RyuJIT


```
|                                     Method |         Mean |      Error |      StdDev |  Ratio | RatioSD |  Gen 0 | Allocated |
|------------------------------------------- |-------------:|-----------:|------------:|-------:|--------:|-------:|----------:|
|                              EnumIsDefined |    35.902 ns |  0.6696 ns |   0.6263 ns |   1.00 |    0.00 |      - |         - |
|                        ExtensionsIsDefined |     3.668 ns |  0.0106 ns |   0.0099 ns |   0.10 |    0.00 |      - |         - |
| EnumIsDefinedNameDisplayName | 4,209.063 ns | 81.6949 ns | 124.7567 ns | 119.50 |    3.04 | 0.1602 |   1,100 B |
|         ExtensionsIsDefinedNameDisplayName |     2.615 ns |  0.0111 ns |   0.0093 ns |   0.07 |    0.00 |      - |         - |


``` ini

BenchmarkDotNet=v0.13.1, OS=Windows 10.0.22000
11th Gen Intel Core i7-11800H 2.30GHz, 1 CPU, 16 logical and 8 physical cores
  [Host]     : .NET Framework 4.8 (4.8.4470.0), X64 RyuJIT  [AttachedDebugger]
  DefaultJob : .NET Framework 4.8 (4.8.4470.0), X64 RyuJIT


```
|                                Method |         Mean |      Error |     StdDev |  Ratio | RatioSD |  Gen 0 | Allocated |
|-------------------------------------- |-------------:|-----------:|-----------:|-------:|--------:|-------:|----------:|
|                          EnumToString |   231.936 ns |  4.6595 ns |  8.4021 ns |  1.000 |    0.00 | 0.0112 |      72 B |
| EnumToStringDisplayName | 2,457.371 ns | 47.3430 ns | 44.2846 ns | 10.883 |    0.48 | 0.1259 |     835 B |
|                          ToStringFast |     1.464 ns |  0.0549 ns |  0.0513 ns |  0.006 |    0.00 |      - |         - |


``` ini

BenchmarkDotNet=v0.13.1, OS=Windows 10.0.22000
11th Gen Intel Core i7-11800H 2.30GHz, 1 CPU, 16 logical and 8 physical cores
  [Host]     : .NET Framework 4.8 (4.8.4470.0), X64 RyuJIT  [AttachedDebugger]
  DefaultJob : .NET Framework 4.8 (4.8.4470.0), X64 RyuJIT


```
|                                Method |         Mean |      Error |     StdDev | Ratio | RatioSD |  Gen 0 | Allocated |
|-------------------------------------- |-------------:|-----------:|-----------:|------:|--------:|-------:|----------:|
|                          EnumTryParse |   165.700 ns |  3.3072 ns |  5.1489 ns |  1.00 |    0.00 | 0.0200 |     128 B |
|                    ExtensionsTryParse |     5.273 ns |  0.1244 ns |  0.1480 ns |  0.03 |    0.00 |      - |         - |
| EnumTryParseDisplayName | 4,053.959 ns | 80.5383 ns | 75.3355 ns | 24.99 |    1.12 | 0.1602 |   1,100 B |
|         ExtensionsTryParseDisplayName |     3.118 ns |  0.0781 ns |  0.0692 ns |  0.02 |    0.00 |      - |         - |

andrewlock commented on tests/NetEscapades.EnumGenerators.Tests/Snapshots/EnumGeneratorTests.CanGenerateEnumExtensionsWithDisplayName.verified.txt:
```diff
@@ -0,0 +1,151 @@
+            string? name, 
+            bool ignoreCase, 
+            out MyTestNameSpace.MyEnum value, 
+            bool allowMatchingDisplayAttribute = false)
```
I know it's opinionated, but I'd prefer not to use an optional parameter for this methods, and instead to have an additional overload. Which would give us

```
TryParse(string name, out MyEnum value);
TryParse(string name, out MyEnum value, bool ignoreCase);
TryParse(string name, out MyEnum value, bool ignoreCase, bool allowMatchingDisplayAttribute);
```

How does that sound to you?

adamradocz replied:
I'll do it, but can you tell me what's the rational behind it?

andrewlock replied:
Sure, optional parameters are defined in the callee's code rather than in the library itself, which _can_ lead to issues with library maintenance. I don't forsee it being much of an issue in this case, but I've been bitten about it before. [This sonarcube rule covers it a bit too](https://rules.sonarsource.com/csharp/RSPEC-2360).

adamradocz replied:
Done

andrewlock commented on tests/NetEscapades.EnumGenerators.Tests/Snapshots/EnumGeneratorTests.CanGenerateEnumExtensionsWithCustomNamespaceAndName.verified.txt:
```diff
@@ -75,23 +76,7 @@ namespace A.B
             string? name, 
             out MyTestNameSpace.MyEnum value)
-        {
-            switch (name)
-            {
-                case nameof(MyTestNameSpace.MyEnum.First):
-                    value = MyTestNameSpace.MyEnum.First;
-                    return true;
-                case nameof(MyTestNameSpace.MyEnum.Second):
-                    value = MyTestNameSpace.MyEnum.Second;
-                    return true;
-                case { } s when int.TryParse(name, out var val):
-                    value = (MyTestNameSpace.MyEnum)val;
-                    return true;
-                default:
-                    value = default;
-                    return false;
-            }
```
We're no longer special-casing this, I wonder what the impact is on going through the switch-expression with string comparison over the simple switch 🤔

adamradocz replied:
What do you mean by special casting? Is it the int parse?

```
case { } s when int.TryParse(name, out var val):
                    value = (MyTestNameSpace.MyEnum)val;
                    return true;

```
String comparison must be slower, as the simple switch is hashed, so it's almost O(1). That's why the `current.Equals(other, Compare.Ordinal)` is slower in the Spanify branch. 
Or you though to something else?

andrewlock replied:
> What do you mean by special casing

Sorry, I meant _not_ having the switch for the case-sensitive case seems like a regression (exactly what you described). 

I don't really see a reason to not have the separate `TryParse` and `TryParseIgnoreCase` branches, given we can optimise the `TryParse` case?

adamradocz replied:
Optimized the TryParse when the case if not ignored.

andrewlock reviewed (commented):
I think this is looking really good @adamradocz, thanks 🙂 Just a couple of final questions/suggestions!

adamradocz commented:
The optimized TryParse is a little bit faster.
|                 Method |       Mean |     Error |    StdDev | Ratio |  Gen 0 | Allocated |
|----------------------- |-----------:|----------:|----------:|------:|-------:|----------:|
|     ExtensionsTryParse |   7.615 ns | 0.0208 ns | 0.0194 ns | 0.008 |      - |         - |
| New ExtensionsTryParse |   7.097 ns | 0.0226 ns | 0.0212 ns | 0.007 |      - |         - |

adamradocz commented:
@andrewlock I think, the `IsDefined` should have the `allowMatchingDisplayAttribute` parameter in its signature, even though the display attribute is not defined in the `enum`, just to be consistent. What do you think?

andrewlock commented:
> @andrewlock I think, the `IsDefined` should have the `allowMatchingDisplayAttribute` parameter in its signature, even though the display attribute is not defined in the `enum`, just to be consistent. What do you think?

I agree. It would be annoying (for example) if I had a `[Display]` attribute, originally, used the `allowMatchingDisplayAttribute` version, removed the attribute later, and then my code wouldn't compile.

Just a thought - I think we should support the `[Description]` attribute too. I'll do that in a separate PR, but I wonder if we should rename `allowMatchingDisplayAttribute` to `allowMatchingMetadataAttribute` (or something) to cover both cases. I don't really like the name as much, but maybe it's better 🤔

adamradocz commented:
> 

How would you differentiate the two attributes then, when parsing? What if I set the `[Description]` and `[Display]` attributes as well, but I don't want the parser to match the description, only for the display?

adamradocz commented:
`IsDefined` is updated.

The only questioning thing is the `allowMatchingDisplayAttribute` vs `allowMatchingMetadataAttribute`.

FroggieFrog commented:
How about adding an additional property to the attribute and respect the defined order?
Something like this:
```csharp
public class EnumExtensionsAttribute : System.Attribute
{
    //...
    public Type[]? AdditionalMetaData { get; set; }
}

[EnumExtensions(AdditionalMetaData = new[] { typeof(DisplayAttribute), typeof(DescriptionAttribute) })]
public enum MyEnum
{
        [Display(Description ="1st")]
        First = 0,
        [Description("2nd")]
        Second = 1,
        [Description("3rd")]
        [Display(Description = "the third")]
        Third = 2,
}
```
So `MyEnum.Third.ToStringFast()` will result in `"the third"`, because `DisplayAttribute` is defined first in `AdditionalMetaData`.

andrewlock commented:
I'm undecided. Personally, I don't think it's worth fighting with this edge case too much. I would probably just use whichever is defined first. I can't imagine anyone is adding _both_ the `[Description]` attribute _and_ the `[Display]` attribute to their enums! If someone is, and we want to support that scenario, we could add the option later. How does that sound @FroggieFrog @adamradocz?

adamradocz commented:
It's okay with me. I'll update the naming soon.

adamradocz commented:
I renamed the variable. I think this PR is complete.

andrewlock commented:
Sorry for the delay on merging this @adamradocz, and thanks for all your hard work! It's much appreciated 🙂
